domain/user/state/state.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/user"
+	"github.com/juju/juju/domain"
+	domainuser "github.com/juju/juju/domain/user"
+	databaseutils "github.com/juju/juju/internal/database"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// AddUser will add a new user to the database. If the user already exists
+// an error that satisfies usererrors.AlreadyExists will be returned. If the
+// users creator is set and does not exist then an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUser(ctx context.Context, domainUser domainuser.User) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	addUserQuery := "INSERT INTO user (uuid, name, display_name, created_by_uuid, created_at) VALUES (?, ?, ?, ?, ?)"
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		_, err := tx.ExecContext(ctx, addUserQuery,
+			domainUser.UUID,
+			domainUser.Name,
+			domainUser.DisplayName,
+			domainUser.CreatorUUID,
+			domainUser.CreatedAt,
+		)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		if databaseutils.IsErrConstraintUnique(err) {
+			return fmt.Errorf("user %q already exists: %w", domainUser.Name, err)
+		}
+		if databaseutils.IsErrConstraintForeignKey(err) {
+			return fmt.Errorf("user %q creator %q does not exist: %w", domainUser.Name, domainUser.CreatorUUID, err)
+		}
+		return fmt.Errorf("cannot insert user %q: %w", domainUser.Name, err)
+	}
+
+	return nil
+}
+
+// AddUserWithPasswordHash will add a new user to the database with the
+// provided password hash and salt. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. If the users creator
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithPasswordHash(ctx context.Context, domainUser domainuser.User, passwordHash string, salt []byte) error {
+	err := st.AddUser(ctx, domainUser)
```

Comment:
Thanks for implementing this however we are going to need to change this a bit.

For database operations like this we need all of the database operations (insert, update, delete etc) to all happen within the one database transaction.

We want to do it this way as having them in seperate transaction means that if one operation succeeds and another operation fails in different transactions we have no safe way to rollback the database state to a known good place.

Even the act of trying to roll back a transaction that has already been committed could fail.

Your initial thought around  trying to compose the problems into distinct problem domains and functions make a lot of smart sense.

If you would like to keep these seperate concerns in distinct logic blocks but the same transaction the best way to do this would be something like the following.

```go
func addUser(ctx, tx, domainUser) error {
    addUserQuery := "INSERT INTO user (uuid, name, display_name, created_by_uuid, created_at) VALUES (?, ?, ?, ?, ?)"
    _, err := tx.ExecContext(ctx, addUserQuery,
			domainUser.UUID,
			domainUser.Name,
			domainUser.DisplayName,
			domainUser.CreatorUUID,
			domainUser.CreatedAt,
		)
    if databaseutils.IsErrConstraintUnique(err) {
        return fmt.Errorf("%w %q", usererrors.AlreadyExists, domainUser.Name)
    }
    // rest of error checking logic that addUser needs to return our other error types.
    return nil
}

func setPassword(ctx, tx, userId, passwordHash, salt) error {
    // Transaction logic and error handling for setting a users password hash
}

func (st *State) AddUser(ctx context.Context, domainUser domainuser.User) error {
	db, err := st.DB()
	if err != nil {
		return fmt.Errorf("cannot get DB access: %w", err)
	}
        return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
            return addUser(ctx, tx, domainUser)
        }
}

func (st *State) AddUserWithPasswordHash(ctx context.Context, domainUser domainuser.User, passwordHash string, salt []byte) error {
    db, err := st.DB()
	if err != nil {
		return fmt.Errorf("cannot get DB access: %w", err)
	}
    return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
            addUser(ctx, tx, domainUser)
           // setPassword
        }
}
```

Hope my rough outline makes some sense.

Reply:
thx! done